edb/server/protocol/auth_ext/email.py

```diff
@@ -0,0 +1,107 @@
+import asyncio
+import urllib.parse
+import random
+
+from typing import Any, Coroutine
+from edb.server import tenant
+
+from edb.server.config.types import CompositeConfigType
+
+from . import util, ui, smtp
+
+
+async def send_password_reset_email(
+    db: Any,
+    tenant: tenant.Tenant,
+    to_addr: str,
+    reset_url: str,
+    secret_token: str,
+    test_mode: bool,
+):
+    from_addr = util.get_config(db, "ext::auth::SMTPConfig::sender")
+    reset_token_params = urllib.parse.urlencode({"reset_token": secret_token})
+    reset_url = f"{reset_url}?{reset_token_params}"
+    ui_config = util.maybe_get_config(
+        db, "ext::auth::AuthConfig::ui", CompositeConfigType
+    )
+    email_args: dict[str, str]
+    if ui_config is None:
+        email_args = {}
+    else:
+        email_args = dict(
+            app_name=ui_config.app_name,
+            logo_url=ui_config.logo_url,
+            dark_logo_url=ui_config.dark_logo_url,
+            brand_color=ui_config.brand_color,
+        )
+    msg = ui.render_password_reset_email(
+        from_addr=from_addr,
+        to_addr=to_addr,
+        reset_url=reset_url,
+        **email_args,
+    )
+    coro = smtp.send_email(
+        db,
+        msg,
+        sender=from_addr,
+        recipients=to_addr,
+        test_mode=test_mode,
+    )
+    await _protected_send(coro, tenant)
+
+
+async def send_verification_email(
+    db: Any,
+    tenant: tenant.Tenant,
+    to_addr: str,
+    verify_url: str,
+    verification_token: str,
+    provider: str,
+    test_mode: bool,
+):
+    from_addr = util.get_config(db, "ext::auth::SMTPConfig::sender")
+    ui_config = util.maybe_get_config(
+        db, "ext::auth::AuthConfig::ui", CompositeConfigType
+    )
+    verification_token_params = urllib.parse.urlencode(
+        {
+            "verification_token": verification_token,
+            "provider": provider,
+        }
+    )
+    verify_url = f"{verify_url}?{verification_token_params}"
+    email_args: dict[str, str]
+    if ui_config is None:
+        email_args = {}
+    else:
+        email_args = dict(
+            app_name=ui_config.app_name,
+            logo_url=ui_config.logo_url,
+            dark_logo_url=ui_config.dark_logo_url,
+            brand_color=ui_config.brand_color,
+        )
+    msg = ui.render_verification_email(
+        from_addr=from_addr,
+        to_addr=to_addr,
+        verify_url=verify_url,
+        **email_args,
+    )
+    coro = smtp.send_email(
+        db,
+        msg,
+        sender=from_addr,
+        recipients=to_addr,
+        test_mode=test_mode,
+    )
+    await _protected_send(coro, tenant)
+
+
+async def _protected_send(
```

Comment:
I don't 100% understand what's going on here. The idea is to execute `coro` but take a random amount of time before returning?
(What's the timing attack we're worried about here?)

The reason I came here to comment was to say that the create_task usage is dangerous because it doesn't save a reference, but then I noticed it was `tenant.create_task`, so clearly fine.

Reply:
I took this code, pretty much verbatim from: https://github.com/edgedb/edgedb/pull/6211/files#diff-ceb3a40c7d3d1fc8a40b20a415f88b76616edc12d2d37ddd46c99b5e94cff9a2R375-R395

If there is a better way to abstract this, I'm definitely open for suggestion, but I'm not sure what the original concern was w.r.t. timing attacks other than making the request take slightly longer to make it a bit more expensive to make the request.

@fantix any thoughts about @msullivan 's question here?

Reply:
I think Yury's idea was to mask the cases where e.g. the email address doesn't exist, or it fails fast to send the email, so that the attackers cannot guess what's happening on the server side from how long it runs, thus protecting the email addresses that do exist from spam or phishing (like, a potential attacker could run a test with a public email list against the server without this sleep, get a list of registered email addresses, and phish saying something as if it's actually from this server). But after reviewing the code, I think the sleep should be done earlier, and `NoIdentityFound` must not be raised in such APIs.

Reply:
@fantix thanks for the reminder on not leaking whether an email is registered!

Looks like we potentially raising `NoIdentityFound` in two places:
1. `resend-verification-email`: if we couldn't find the `email`.
2. `send_reset_email`: if we couldn't find the `email`.
3. `verify`: if we couldn't find the `email` referenced in the verification token.

I think we can and should trust the `email` within the token, so it still seems appropriate to send a useful message here. It _should_ be extremely unlikely to get into this state normally, but if you delete the `LocalIdentity` and `EmailPasswordFactor` for a user who has not verified their email, they would hit this case.

---

What do you think of the `send_fake_email` approach [here](https://github.com/edgedb/edgedb/pull/6324/commits/5e40b54fb771bfba4c5022b5d45bf201c2aeb173) rather than moving the sleep?

Reply:
Yeah, that looks good to me 👍